### PR TITLE
`Adaptive learning`: Add weight to links between competencies and learning objects

### PR DESCRIPTION
**Pull request by JohannesStoehr**
#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/development-process.html#naming-conventions-for-github-pull-requests).

#### Server
- [x] **Important**: I implemented the changes with a [very good performance](https://docs.artemis.cit.tum.de/dev/guidelines/performance/) and prevented too many (unnecessary) and too complex database calls.
- [x] I **strictly** followed the principle of **data economy** for all database calls.
- [x] I **strictly** followed the [server coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/).
- [x] I added multiple integration tests (Spring) related to the features (with a high test coverage).
- [x] I documented the Java code using JavaDoc style.

#### Client
- [x] **Important**: I implemented the changes with a very good performance, prevented too many (unnecessary) REST calls and made sure the UI is responsive, even with large data (e.g. using paging).
- [x] I **strictly** followed the principle of **data economy** for all client-server REST calls.
- [x] I **strictly** followed the [client coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client/).
- [x] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-tests/).
- [x] I documented the TypeScript code using JSDoc style.

### Motivation and Context

#9446

### Description

Add a column to the connection between exercises/lecture units (learning objects) and competencies that represents a weight or importance of this learning object for the content of the competency. To reflect this in Java the many-to-many relation is now two many to one relations with a new CompetencyExercise/LectureUnitLink between the learning object and the competency.
The weight is currently always set to 1, it will be adjustable in a follow up.

The table in the competency form is removed for now and will be redone by @ardakaraman0

### Steps for Testing

Prerequisites:
- 1 Instructor
- 2 competencies/prerequisites

1. Log in to Artemis
2. Check that creating/editing/importing an exercise/lecture unit still works in the following situations:
   - Linking a competency
   - Unlinking a competency
   - Changing the linked competencies

[](https://artemis-test1.artemis.cit.tum.de)
[](https://artemis-test2.artemis.cit.tum.de)
[](https://artemis-test3.artemis.cit.tum.de)
[](https://artemis-test4.artemis.cit.tum.de)
[](https://artemis-test5.artemis.cit.tum.de)
[](https://artemis-test6.artemis.cit.tum.de)
[](https://artemis-test9.artemis.cit.tum.de)

#### Performance Review

- [x] I (as a reviewer) confirm that the client changes (in particular related to REST calls and UI responsiveness) are implemented with a very good performance even for very large courses with more than 2000 students.
- [x] I (as a reviewer) confirm that the server changes (in particular related to database calls) are implemented with a very good performance even for very large courses with more than 2000 students.
#### Code Review
- [x] Code Review 1
- [x] Code Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2
#### Performance Tests
- [x] Test 1
- [x] Test 2

#### Server

| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|---------------------:|
| CompetencyExerciseLink.java | 93% | ✅ |
| CompetencyLearningObjectLink.java | 100% | ✅ |
| CompetencyLectureUnitLink.java | 93% | ✅ |

### CODE REVIEW

File: src/test/java/de/tum/cit/aet/artemis/programming/icl/ProgrammingExerciseLocalVCLocalCIIntegrationTest.java
Comment (pzdr7) [inline src/test/java/de/tum/cit/aet/artemis/programming/icl/ProgrammingExerciseLocalVCLocalCIIntegrationTest.java:298]: Does leaving in the course lead to a problem?

**Review by pzdr7 — COMMENTED** [review]
Some comments on the code. Since the code has changed since I started reviewing, some of them might be outdated. In that case, just dismiss them

**Review by ole-ve — APPROVED** [review]
Tested on TS1, works as expected. Nice work

File: src/main/webapp/app/course/competencies/competencies-popover/competencies-popover.component.ts
Comment (JohannesStoehr) [reply src/main/webapp/app/course/competencies/competencies-popover/competencies-popover.component.ts:17]: No, but I would consider this out of scope for this PR. Especially since we are already over 1500 new/changed lined of code and this would add hundreds more

File: src/test/java/de/tum/cit/aet/artemis/programming/icl/ProgrammingExerciseLocalVCLocalCIIntegrationTest.java
Comment (JohannesStoehr) [reply src/test/java/de/tum/cit/aet/artemis/programming/icl/ProgrammingExerciseLocalVCLocalCIIntegrationTest.java:298]: This leads to problems since the course again contains the competency and we get circular dependencies

File: src/main/webapp/app/entities/competency.model.ts
Comment (JohannesWt) [inline src/main/webapp/app/entities/competency.model.ts:87]: can there be a CompetencylearningObjectLink without a competency?

File: src/main/webapp/app/entities/competency.model.ts
Comment (JohannesWt) [inline src/main/webapp/app/entities/competency.model.ts:97]: same here for exercise

File: src/main/webapp/app/entities/competency.model.ts
Comment (JohannesWt) [inline src/main/webapp/app/entities/competency.model.ts:106]: and here for lectureUnit

**Review by JohannesWt — CHANGES_REQUESTED** [review]
Just a few small comments/ideas

File: src/main/webapp/app/course/competencies/forms/competency/competency-form.component.ts
Comment (JohannesStoehr) [reply src/main/webapp/app/course/competencies/forms/competency/competency-form.component.ts:30]: Sadly the formData above cannot be easily be converted to the new signal input, since it is also referenced in the super class with a more generic type. 
Since this would be more effort to refactor this component, I will not do it in this PR.

File: src/main/webapp/app/entities/competency.model.ts
Comment (JohannesStoehr) [reply src/main/webapp/app/entities/competency.model.ts:87]: Yes, since we sadly do not use DTOs for the exercise and competency endpoints yet, a competency might have its links stored but these links do not contain the competency again to avoid circular dependencies.

**Review by SimonEntholzer — COMMENTED** [review]
When importing an exercise, the competencies should not be imported as well? For me I have to newly select them during importing

**Review by sachmii — DISMISSED** [review]
Tested on TS4, works fine.

**Review by HawKhiem — COMMENTED** [review]
Tested on TS4 during the testing session. Linking and unlinking works fine, the changed competency also displayed as expected. One thing I do notice, when I tried to import an exercise with some linked competencies into a different course with the same competencies as the first course, all the linked competencies dissappear. Is this behaviour intended?

Screenshot 2024-10-24 150340
Screenshot 2024-10-24 150557

**Review by Malekos74 — DISMISSED** [review]
### Server tested on: Server 4
Works as expected

File: src/main/webapp/app/course/competencies/competencies-popover/competencies-popover.component.html
Comment (florian-glombik) [inline src/main/webapp/app/course/competencies/competencies-popover/competencies-popover.component.html:8]: Would it make sense to wrap the link in an if instead of the optional chaining?
If the title is not there the link will most likely look quite broken

File: src/main/webapp/app/overview/course-competencies/course-competencies-details.component.html
Comment (florian-glombik) [inline src/main/webapp/app/overview/course-competencies/course-competencies-details.component.html:72]: Can we get rid of the errors here?

> Type LectureUnit is not assignable to type T 

If there is no other way to fix the type a directive similar to `exercise-detail.directive.ts` should solve the issue

File: src/main/webapp/app/overview/course-competencies/course-competencies-details.component.ts
Comment (florian-glombik) [inline src/main/webapp/app/overview/course-competencies/course-competencies-details.component.ts:128]: Could we please define this default value in a variable with a description why we choose 0.5?

File: src/main/webapp/app/shared/competency-selection/competency-selection.component.ts
Comment (florian-glombik) [inline src/main/webapp/app/shared/competency-selection/competency-selection.component.ts:92]: The variable with 0.5 should be reused here

**Review by florian-glombik — CHANGES_REQUESTED** [review]
Reviewed the client code so far (74 / 157), I will have a look at the server code tomorrow

File: src/main/webapp/app/overview/course-competencies/course-competencies-details.component.html
Comment (JohannesStoehr) [reply src/main/webapp/app/overview/course-competencies/course-competencies-details.component.html:72]: They were also present before this PR. We use some kind of polymorphism in a common super class that I do not want to touch in this PR

File: src/main/webapp/app/overview/course-competencies/course-competencies-details.component.ts
Comment (JohannesStoehr) [reply src/main/webapp/app/overview/course-competencies/course-competencies-details.component.ts:128]: Done in #9564

**Review by sachmii — DISMISSED** [review]
Re-approve, tested on TS3.

File: src/main/java/de/tum/cit/aet/artemis/atlas/domain/competency/CompetencyLearningObjectLink.java
Comment (florian-glombik) [inline src/main/java/de/tum/cit/aet/artemis/atlas/domain/competency/CompetencyLearningObjectLink.java:44]: Isn't it best practice to define the constructors before the getters and setters?

**Review by HawKhiem — APPROVED** [review]
Tested on TS2. Reapprove

**Review by florian-glombik — DISMISSED** [review]
Approve code

The comment about the ordering of the constructor does not need to be addressed within this PR due to the size of the PR and the effort of re-testing and gathering new reviews :)

**Review by Malekos74 — DISMISSED** [review]
### Server tested on: Server 2
Works as expected

**Review by pzdr7 — APPROVED** [review]
Code

**Review by JohannesWt — DISMISSED** [review]
Code

**Review by Cathy0123456789 — COMMENTED** [review]
Tested on TS3.
Linking / unlinking / changing linked competencies to exercises and lecture units worked. However I have three small questions:
- Linked competencies are shown next to a lecture unit, should it also be visible anywhere within the exercises?
- I was not able to click onto a Prerequisite in the Competencies tab, neither when no lecture unit was linked to it, nor when it was linked (not sure if relevant to this specific PR?)
- When editing the linked competencies of a lecture unit / exercise, sometimes the checkbox next to competencies, which I had linked already earlier, were checked, sometimes not. Maybe this is not a problem of this PR though, as I got "Internal Server Error"s off and on when editing the linked competencies

**Review by pzdr7 — APPROVED** [review]
Reapprove after merge

**Review by JohannesWt — DISMISSED** [review]
Reapprove after merge conflict

**Review by florian-glombik — APPROVED** [review]
Re-approve code

**Review by zagemello — APPROVED** [review]
Was able to link and unlink competencies when importing and creating. Tested on TS5

**Review by sachmii — APPROVED** [review]

**Review by Malekos74 — APPROVED** [review]
Merge Conflict reapprove

**Review by MaximilianAnzinger — DISMISSED** [review]
Maintainer approved

**Review by sachmii — APPROVED** [review]

**Pull request merged**
